processing/src/main/java/org/apache/druid/segment/join/JoinableFactoryWrapper.java

```diff
@@ -276,28 +272,31 @@ static Optional<Filter> convertJoinToFilter(
         && clause.getCondition().getEquiConditions().size() > 0) {
       final List<Filter> filters = new ArrayList<>();
       int numValues = maxNumFilterValues;
+      boolean dropClause = true;
 
       for (final Equality condition : clause.getCondition().getEquiConditions()) {
         final String leftColumn = condition.getLeftExpr().getBindingIfIdentifier();
 
         if (leftColumn == null) {
-          return Optional.empty();
+          return new NonnullPair<>(Optional.empty(), false);
         }
 
-        final Optional<Set<String>> columnValuesForFilter =
-            clause.getJoinable().getNonNullColumnValuesIfAllUnique(condition.getRightColumn(), numValues);
+        Joinable.ColumnValuesWithUniqueFlag columnValuesWithUniqueFlag = clause.getJoinable().getNonNullColumnValues(condition.getRightColumn(), numValues);
+        if (columnValuesWithUniqueFlag.getColumnValues().isEmpty()) {
+          dropClause = false;
+          continue;
+        }
 
-        if (columnValuesForFilter.isPresent()) {
-          numValues -= columnValuesForFilter.get().size();
-          filters.add(Filters.toFilter(new InDimFilter(leftColumn, columnValuesForFilter.get())));
-        } else {
-          return Optional.empty();
+        numValues -= columnValuesWithUniqueFlag.getColumnValues().size();
+        filters.add(Filters.toFilter(new InDimFilter(leftColumn, columnValuesWithUniqueFlag.getColumnValues())));
+        if (!columnValuesWithUniqueFlag.isAllUnique()) {
+          dropClause = false;
         }
       }
 
-      return Optional.of(Filters.and(filters));
+      return new NonnullPair<>(filters.isEmpty() ? Optional.empty() : Optional.of(Filters.and(filters)), dropClause);
```

Comment:
> Will we ever get filters = Optional.empty() and dropClause = true? If so what is the expectation in that case?

Yes, that can happen if the joinableClause has conditions on columns which only have nulls in it. Currently, in such a case I add a `FalseFilter` in place of the joinable clause - does that seem ok? 

> IMO, the NonnullPair is tough to think about here, so a special class would be better. That way it could have some javadocs about the expectations. Its constructor should also do any relevant invariant checks.

Yes, completely agree. 👍  Thanks for the suggestion - I've made a holder instead of a pair and added documentation regarding its semantics.

Reply:
> Yes, that can happen if the joinableClause has conditions on columns which only have nulls in it. Currently, in such a case I add a FalseFilter in place of the joinable clause - does that seem ok?

It makes sense to me. If the behavior is described in javadocs for the new holder class, that's even better.

Reply:
:+1: thanks for suggestion - added to the javadocs for holder class